File: src/zros_cpp/impl/service_clients_impl.cc
//
// Created by pairs on 7/18/18.
//
#include "service_clients_impl.h"
#include <sspdlog/sspdlog.h>

namespace zros {

    GrpcStub::GrpcStub(const std::string &address):
            RpcStub(address),
            node_stub_(std::make_shared<NodeStub >(createNodeStub(address))) {
    }

    std::shared_ptr<zros_rpc::ServiceResponse> GrpcStub::call(const std::string &service_name,
                                                              const std::string &content, const std::string &cli_info,
                                                              int timeout_mseconds) {
        auto service_response = std::make_shared<zros_rpc::ServiceResponse>();
        auto service_request = std::make_shared<zros_rpc::ServiceRequest>();
        service_request->set_service_name(service_name);
        service_request->set_request_data(content);
        service_request->mutable_cli_node_info()->set_address(cli_info);

        auto context = std::make_shared<grpc::ClientContext>();
        if (timeout_mseconds >= 0) {
            context->set_deadline(std::chrono::system_clock::now() + std::chrono::milliseconds(timeout_mseconds));
        }
        auto status = (*node_stub_)->InvokeService(context.get(), *service_request, service_response.get());
        if (!status.ok()) {
            service_response->mutable_status()->set_code(zros_rpc::Status_Code_UNKNOWN);
            service_response->mutable_status()->set_details(std::string("GRPC INNER ERROR: ") + status.error_message());
        }
        return service_response;
    }

    NodeStub GrpcStub::createNodeStub(const std::string &address) {
        grpc::ChannelArguments args;
        args.SetInt(GRPC_ARG_MAX_RECONNECT_BACKOFF_MS, 5*1000);
        return zros_rpc::ServiceRPC::NewStub(
                grpc::CreateCustomChannel(address, grpc::InsecureChannelCredentials(), args)
        );
    }

    void ServiceClientsImpl::registerServiceClient(const zros_rpc::ServiceServerInfo *serverInfo, zros_rpc::Status *status) {
        std::string service_name = serverInfo->service_name();
        std::lock_guard<std::mutex> lk1(clients_mutex_);
        if (clients_.find(service_name) == clients_.end()) {
            SSPD_LOG_WARNING << "please check registerServiceClient logical error " << service_name;
            status->set_code(status->NOT_FOUND);
            status->set_details(service_name + " client not register on");
            return;
        }
        clients_[service_name]->set_ready(true);
        std::lock_guard<std::mutex> lk2(services_mutex_);
        services_address_[service_name] = serverInfo->physical_node_info().real_address();
        status->set_code(status->OK);
        SSPD_LOG_INFO << "register client : " << service_name << "on address: " << serverInfo->physical_node_info().real_address() << "success";
    }

    void ServiceClientsImpl::unregisterServiceClient(const zros_rpc::ServiceServerInfo *serverInfo,
                                                     zros_rpc::Status *status) {
        // todo
    }

    bool ServiceClientsImpl::unregisterServiceClient(const std::string &service_name) {
        // todo
        return false;
    }

    std::shared_ptr<zros_rpc::ServiceResponse>
    ServiceClientsImpl::call(const std::string &service_name, const std::string &content, const std::string &cli_info,
                             int timeout_mseconds) {
        std::lock_guard<std::mutex> lk (clients_mutex_);
        if (clients_.find(service_name) != clients_.end()) {
            SSPD_LOG_ERROR << "please check  if register this client ";
            return nullptr;
        }
        // already ready
        auto client = clients_[service_name];
        auto stub = get_rpc_stub(service_name);
        if (stub == nullptr) {
            SSPD_LOG_WARNING << "stub is nullptr ";
            auto tmp_res = std::make_shared<zros_rpc::ServiceResponse>();
            tmp_res->mutable_status()->set_code(zros_rpc::Status_Code_NOT_FOUND);
            tmp_res->mutable_status()->set_details("service call stub not found, probably waiting for establishing yet.");
            return tmp_res;
        }
        return stub->call(service_name, content, cli_info, timeout_mseconds);
    }

    bool ServiceClientsImpl::registerClient(const std::shared_ptr<IServiceClient> client) {
        std::lock_guard<std::mutex> lk(clients_mutex_);
        if (clients_.find(client->get_service_name()) != clients_.end()) {
            SSPD_LOG_WARNING << client->get_service_name() << " already register";
            return false;
        }
        return true;
    }

    std::shared_ptr<RpcStub> ServiceClientsImpl::get_rpc_stub(const std::string &service_name) {
        std::lock_guard<std::mutex> lk1(services_mutex_);
        std::lock_guard<std::mutex> lk2(stubs_mutex_);
        if(services_address_.find(service_name) == services_address_.end()) {
            SSPD_LOG_ERROR << "get rpc stub failed: there is no this " << service_name << "service client";
            return nullptr;
        }
        std::string address = services_address_[service_name];
        auto stub = stubs_.find(address);
        if (stub == stubs_.end()) {
            // first use
            stubs_[address] = std::make_shared<GrpcStub>(address);
        }
        return stubs_[address];
    }

}   // namespace zros

File: src/zros_python/impl/service_clients_impl.py
# -*- coding: utf-8 -*-
""" impl service client management

data: 2018-08-01
author: pairs
"""

import abc
import logging
import threading
import grpc
import zros_python.pb.zros_pb2 as zpb2
import zros_python.pb.zros_pb2_grpc as zpbg2
import traceback

logger = logging.getLogger(__name__)


class RpcStub(object):
    __metaclass__ = abc.ABCMeta
    """
    
    """
    def __init__(self):
        pass

    @abc.abstractmethod
    def call(self, service_name, content, cli_info, timeout):
        pass


class GrpcStub(RpcStub):
    def __init__(self, address):
        super(GrpcStub, self).__init__()
        channel = grpc.insecure_channel(address)
        self._stub = zpbg2.ServiceRPCStub(channel)

    def call(self, service_name, content, cli_info, timeout):
        request = zpb2.ServiceRequest()
        response = zpb2.ServiceResponse()
        request.service_name = service_name
        request.request_data = content
        try:
            return self._stub.InvokeService(request)
        except Exception:
            logger.error(u'GRPC INNER ERROR' + traceback.format_exc())
            response.status.code = zpb2.Status.UNKNOWN
            response.status.details = u'GRPC INNER ERROR' + traceback.format_exc()
            return response


class ServiceClientsImpl(object):
    """

    """
    def __init__(self):
        self._clients_mutex = threading.Lock()
        self._services_mutex = threading.Lock()
        self._stubs_mutex = threading.Lock()
        self._clients = {}
        self._services_address = {}
        self._stubs = {}

    def register_client(self, client):
        service_name = client.get_service_name()
        with self._clients_mutex:
            if service_name in self._clients:
                logger.warning(u'service client %s already register', service_name)
            else:
                self._clients[service_name] = client

    def unregister_client(self, service_name):
        # todo
        pass

    def call(self, service_name, content, cli_info, timeout):
        response = zpb2.ServiceResponse()
        with self._clients_mutex:
            if service_name not in self._clients:
                logger.error(u'please check if register this client %s', service_name)
                response.status.code = zpb2.Status.UNKNOWN
                response.status.details = u'please check if register this client ' + service_name
                return response

        stub = self._get_rpc_stub(service_name)
        if stub is None:
            logger.warning(u'stub is None')
            response.status.code = zpb2.Status.NOT_FOUND
            response.status.details = u'service call stub not found, probably waiting for establishing yet.'
            return response
        return stub.call(service_name, content, cli_info, timeout)

    def register_service_client(self, server_info):
        service_name = server_info.service_name
        status = zpb2.Status()
        with self._clients_mutex:
            if service_name not in self._clients:
                logger.warning(u'please check register service client logical error %s', service_name)
                status.code = zpb2.Status.NOT_FOUND
                status.details = service_name + u' client not register on'
                return status
            self._clients[service_name].set_ready(True)

        with self._services_mutex:
            self._services_address[service_name] = server_info.physical_node_info.real_address
            status.code = zpb2.Status.OK
            logger.info(u'register client: %s on address %s success', service_name,
                        server_info.physical_node_info.real_address)
            return status

    def unregister_service_client(self, server_info):
        service_name = server_info.service_name
        status = zpb2.Status()
        logger.info(u'unregister service client %s ', service_name)
        # todo
        return status

    def _get_rpc_stub(self, service_name):
        with self._services_mutex:
            with self._stubs_mutex:
                if service_name not in self._services_address:
                    logger.error(u'get rpc stub failed: there is no this %s service client', service_name)
                    return None
                address = self._services_address[service_name]
                if address not in self._stubs:
                    self._stubs[address] = GrpcStub(address)
                return self._stubs[address]
